CDAP-9283 Background upgrade thread

gokulavasan opened the pull request:
JIRA : https://issues.cask.co/browse/CDAP-9283

Build : https://builds.cask.co/browse/CDAP-RUT864

gokulavasan committed:
CDAP-9283 AppVersion upgrade thread, service and handler. Use the appversion upgrade key to avoid unnecessary operations in DatasetBasedStreamSizeScheduleStore Background upgrade for DatasetBasedTimeScheduleStore

anew commented:
It looks like this includes all the changes from https://github.com/caskdata/cdap/pull/8486?

anew commented:
I am gonna review only the second commit. Please let me know if I understood that correctly.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -73,24 +79,48 @@
+  public synchronized void shutdown() throws Exception {
+    if (executorService != null) {
+      executorService.shutdown();
+      executorService.awaitTermination(1, TimeUnit.SECONDS);
```
what's the point of waiting? Isn't it a daemon thread?

gokulavasan replied:
Yes, it is a daemon thread. Removed.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -183,8 +213,8 @@ public void updateLastRun(ProgramId programId, SchedulableProgramType programTyp
     updateTable(programId, programType, scheduleName,
-                new byte[][]{ LAST_RUN_SIZE_COL, LAST_RUN_TS_COL },
-                new byte[][]{ Bytes.toBytes(newLastRunSize), Bytes.toBytes(newLastRunTs) },
+                new byte[][]{LAST_RUN_SIZE_COL, LAST_RUN_TS_COL},
```
why remove spaces?

gokulavasan replied:
Mistake, reverted.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -353,22 +385,49 @@ public void apply() throws Exception {
+
+    // Wait until the store is initialized
+    while (!storeInitialized.get()) {
+      TimeUnit.MINUTES.sleep(1);
```
sleep a whole minute? Do you expect it to take that long to initialize the store?

gokulavasan replied:
Reduced it to 10 seconds. Should it be even tighter?

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -353,22 +385,49 @@ public void apply() throws Exception {
-        }
-      });
+    throws InterruptedException, IOException, DatasetManagementException {
+    LOG.info("Starting upgrade of DatasetBasedStreamSizeScheduleStore.");
```
maybe this should only be logged after the sleep loop below.

gokulavasan replied:
Moved.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -353,22 +385,49 @@ public void apply() throws Exception {
-  private void upgradeVersionKeys() {
+  // Returns true if the upgrade flag is set. Upgrade could have completed earlier than this since this flag is
+  // updated asynchronously.
+  public boolean isUpgradeComplete() {
```
why not use this method everywhere else? Makes the code a lot more readable

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -353,22 +385,49 @@ public void apply() throws Exception {
+            }
+          });
+      } catch (TransactionFailureException e) {
+        LOG.debug("Upgrade step faced Transaction failure. Will retry operation after some delay.", e);
```
there are various possibilities:
- conflict: in this case you want to retry right away. 
- timeout: you want to try again with longer timeout? Or with fewer records?
- other errors: thse would be unrelated to transactions, and I think not retryable. They should be logged as errors not debug.

Also, can other operations fail with conflicts while the upgrade is running)?

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AppVersionUpgradeService.java:
```diff
@@ -0,0 +1,79 @@
+  AppVersionUpgradeService(DatasetBasedStreamSizeScheduleStore streamSizeScheduleStore) {
+    this.streamSizeScheduleStore = streamSizeScheduleStore;
+    this.executorService = Executors.newFixedThreadPool(
+      3, Threads.createDaemonThreadFactory("app-version-upgrade-thread-%d"));
```
why three threads? If you only submit one runnable?

gokulavasan replied:
We will be adding two more threads for DatasetBasedTimeScheduleStore upgrade and AppMetadataStore upgrade. I have reduced it to 1 for now, will bump it up in subsequent commits.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -339,22 +385,49 @@ public void apply() throws Exception {
+            public void apply() {
+              if (upgradeVersionKeys(table)) {
+                // Upgrade is complete. Mark that app version upgrade is complete in the table.
+                table.put(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'), APP_VERSION_UPGRADE_KEY);
```
perhaps good idea to use the current version as value. That way, when we have an upgrade next time, we can distinguish.

anew replied:
don't you also want to read this at the very beginning? If it is already set, then there is no need to try and upgrade more.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -350,25 +385,98 @@ public void apply() throws Exception {
+          if (currMaxRows > 1) {
+            maxRows.decrementAndGet();
+          }
+          sleepTimeInMins.set(0);
```
Why no need to sleep?

gokulavasan replied:
Since the exception was due to transaction timeout and hence we retry with smaller number of rows.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -350,25 +385,98 @@ public void apply() throws Exception {
+          }
+          sleepTimeInMins.set(0);
+          LOG.debug("Upgrade step faced a Transaction Timeout exception. " +
+                      "Reducing the number of max rows to : {} and retrying the operation now.", maxRows.get(), e);
```
Log message seems confusing if `currMaxRows <= 1`.

gokulavasan replied:
Hmm, that case won't occur?

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -350,25 +385,98 @@ public void apply() throws Exception {
+          int currMaxRows = maxRows.get();
+          if (currMaxRows > 1) {
+            maxRows.decrementAndGet();
+          }
```
if `currMaxRows == 0`, still need to retry?

gokulavasan replied:
currMaxRows just indicates the max number of rows to scan.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -350,25 +385,98 @@ public void apply() throws Exception {
+  private boolean upgradeVersionKeys(Table table, int maxNumRows) {
+    int numRowsUpgraded = 0;
     Joiner joiner = Joiner.on(":");
     try (Scanner scan = getScannerWithPrefix(KEY_PREFIX)) {
```
The scanner here always starts scanning from `KEY_PREFIX`. Is it assumed that the keys in the older format will always precede upgraded keys? How is this guaranteed?

gokulavasan replied:
No there is no assumption. We will keep scanning until we hit an old format row. Do you think this is a problem?

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -89,11 +106,23 @@ public void initialize(ClassLoadHelper loadHelper, SchedulerSignaler schedSignal
   }
 
+  @Override
+  public void shutdown() {
```
why is `DatasetBasedStreamSizeScheduleStore.shutdown()` synchronized but not necessary here?

gokulavasan replied:
I guess it is because DatasetBasedStreamSizeScheduleStore is a Singleton while this one isn't. I don't know why that is the case though.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -173,7 +202,7 @@ private void executeDelete(final TriggerKey triggerKey) {
           @Override
           public void apply() throws Exception {
-            if (versionLessTriggerKey != null && readTrigger(versionLessTriggerKey) != null) {
+            if (!isUpgradeComplete() && versionLessTriggerKey != null && readTrigger(versionLessTriggerKey) != null) {
```
only need to call `removeAppVersion(triggerKey)` if `!isUpgradeComplete()`, so that no need of `!isUpgradeComplete() &&` here

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -350,25 +385,98 @@ public void apply() throws Exception {
+          @Override
+          public void apply() throws Exception {
+            byte[] result = table.get(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'));
+            if (result != null) {
```
why not compare `result` with `Bytes.toBytes(ProjectInfo.getVersion().toString())`?

poornachandra replied:
👍

gokulavasan replied:
Why should it be compared with current version? What will happen in the next version of CDAP?

poornachandra replied:
Not the current version, let's use version 4.1.1

ecfm replied:
Isn't `ProjectInfo.getVersion()` become the next version when the next version of CDAP runs upgrade? My point is it possible that upgrade has already been run for the table but now we are upgrading it again to a newer version, and `result != null` is still true before we run upgrade again

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -383,22 +412,82 @@ public void apply() throws Exception {
+          @Override
+          public void apply() throws Exception {
+            byte[] result = table.get(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'));
+            if (result != null) {
```
why not compare result with `Bytes.toBytes(ProjectInfo.getVersion().toString())`?

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -383,22 +412,82 @@ public void apply() throws Exception {
+    while (!isUpgradeComplete()) {
+      sleepTimeInMins.set(1);
+      try {
+        factory.createExecutor(ImmutableList.of((TransactionAware) table))
```
how can `upgrade()` be executed in the background with table locked to complete `upgradeJobs` and `upgradeTriggers` in a single transaction?

gokulavasan replied:
We store the data in columns of a single row. And since by default the transaction conflict is set to ROW level, we can't do much. I will open a JIRA to migrate this data to rows since having 10k+ columns is not good.

poornachandra replied:
JIRA filed - https://issues.cask.co/browse/CDAP-9401

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:
```diff
@@ -802,6 +806,24 @@ public WorkflowToken getWorkflowToken(ProgramId workflowId, String workflowRunId
+  public boolean isUpgradeComplete(byte[] key) {
+    MDSKey.Builder keyBuilder = new MDSKey.Builder();
+    keyBuilder.add(key);
+    return exists(keyBuilder.build());
```
why not compare the content of the row with `ProjectInfo.getVersion().toString()`?

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -62,6 +71,7 @@
 
   private static final Gson GSON = new Gson();
   private static final String KEY_PREFIX = "streamSizeSchedule";
+  private static final byte[] APP_VERSION_UPGRADE_KEY = Bytes.toBytes("upgrade.streamsize.appversion.complete");
```
Since we store the version as the value, it would be better to call this `version.streamsize.schedule`. Also having `version` in the beginning of the key will lead to some uniformity.

gokulavasan replied:
Changed.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -350,25 +385,98 @@ public void apply() throws Exception {
-      while ((next = scan.next()) != null) {
+      // Upgrade only 10 rows in one transaction to reduce the probability of conflicts with regular Store operations.
+      while (((next = scan.next()) != null) && (numRowsUpgraded < maxNumRows)) {
         if (isInvalidRow(next)) {
```
If all the rows are invalid for this batch then `numRowsUpgraded` will be zero even though there are records to be upgraded, right?

gokulavasan replied:
There is no batching. We just keep scanning until we exhaust all the rows or if we hit the max number of rows to be upgraded.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -57,27 +62,39 @@
  */
 public class DatasetBasedTimeScheduleStore extends RAMJobStore {
 
+  private static final byte[] APP_VERSION_UPGRADE_KEY = Bytes.toBytes("upgrade.timeschedule.appversion.complete");
```
See my earlier note about naming

poornachandra reviewed (commented):
@gokulavasan I have a couple of comments for now, still going through other files.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/UpgradeHttpHandler.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  @GET
+  @Path("/status")
```
This REST API needs to be documented. Can be done later, just a reminder.

sreevatsanraman replied:
Added a JIRA to doc this. https://issues.cask.co/browse/CDAP-9400

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -350,25 +385,99 @@ public void apply() throws Exception {
+      return;
+    }
+
+    final AtomicInteger maxRows = new AtomicInteger(10);
```
Let's increase max rows to 1000

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -57,27 +62,39 @@
   private static final Logger LOG = LoggerFactory.getLogger(DatasetBasedTimeScheduleStore.class);
   private static final byte[] JOB_KEY = Bytes.toBytes("jobs");
   private static final byte[] TRIGGER_KEY = Bytes.toBytes("trigger");
+  private static volatile boolean storeInitialized = false;
```
Schedule table below gets initialized for every instance of DatasetBasedTimeScheduleStore, shouldn't `storeInitialized` also be set for every instance?

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -89,11 +106,23 @@ public void initialize(ClassLoadHelper loadHelper, SchedulerSignaler schedSignal
       readSchedulesFromPersistentStore();
+      executorService = Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory(
+        "time-schedulestore-appver-upgrade-checker"));
+      executorService.submit(new UpgradeCompleteCheck(APP_VERSION_UPGRADE_KEY, factory, table, appVerUpgrade));
```
I think it would be better to check for the upgrade version right after store is initialized, so that in subsequent restarts of CDAP after the upgrade is done we don't have to spawn the check thread.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/UpgradeCompleteCheck.java:
```diff
@@ -0,0 +1,76 @@
+            @Override
+            public void apply() throws Exception {
+              Row row = table.get(upgradeKey);
+              if (!row.isEmpty()) {
```
Would be good to check for version here

gokulavasan replied:
What should be the logic of version check? Should I check for version >= 3.5.4?

poornachandra replied:
It should be >= 4.1.1, right?

gokulavasan replied:
Sorry, I meant >= 4.1.1.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -221,16 +253,20 @@ public void apply() throws Exception {
           }
 
           table.delete(Bytes.toBytes(rowKey));
+
```
remove empty line

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AppVersionUpgradeService.java:
```diff
@@ -0,0 +1,104 @@
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting AppVersionUpgradeService.");
+    executorService.submit(new Runnable() {
```
It would be good to check if upgrade is done before submitting upgrade jobs

poornachandra reviewed (commented):
@gokulavasan Few more comments

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/AbstractSchedulerService.java:
```diff
@@ -229,6 +233,18 @@ public ScheduleState scheduleState(ProgramId program, SchedulableProgramType pro
   }
 
+  @Nullable
+  public static String removeAppVersion(String scheduleId) {
```
Looks like this method was already part of https://github.com/caskdata/cdap/pull/8486

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -73,24 +83,46 @@
    */
   public synchronized void initialize() throws IOException, DatasetManagementException {
+    initializeTable();
+    executorService = Executors.newSingleThreadExecutor(
```
It would be good to check if store has already been upgraded before starting this thread

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -245,22 +290,26 @@ public void apply() throws Exception {
-              if (splits.length != 7) {
+              ProgramId program;
+
+              if (splits.length == 7) {
```
Looks like this snippet of code is also present in https://github.com/caskdata/cdap/pull/8486

gokulavasan replied:
Not sure about this comment. Seems correct to me.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -1072,4 +1161,51 @@ private ApplicationSpecificationWithChangedWorkers(ApplicationSpecification dele
     });
   }
+
+  private static final class UpgradeChecker implements Runnable {
```
Why not use `UpgradeCompleteCheck`?

poornachandra replied:
👍

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -387,13 +496,26 @@ private void upgradeVersionKeys() {
         splitsList.add(3, ApplicationId.DEFAULT_VERSION);
```
Can call the new util method to add the default version

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -383,22 +412,83 @@ public void apply() throws Exception {
-        public void apply() throws Exception {
-          upgradeJobs();
-          upgradeTriggers();
+    try {
```
this part seems to be duplicated with codes in `DatasetBasedStreamSizeScheduleStore` and `DefaultStore`, can add a util method for this

gokulavasan replied:
There are differences between the two since one of them doesn't do multiple operations etc. I will file a JIRA to fix this.

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/UpgradeCompleteCheck.java:
```diff
@@ -0,0 +1,76 @@
+            @Override
+            public void apply() throws Exception {
+              Row row = table.get(upgradeKey);
+              if (!row.isEmpty()) {
```
should check the content in the row

gokulavasan replied:
Added check for that

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -476,7 +566,7 @@ private TriggerStatusV2 addAppVersion(TriggerStatusV2 oldTrigger) {
   }
 
-  private void upgradeJobs() throws Exception {
+  private void upgradeJobs(Table table) {
```
also need to check whether `jobDetail` after upgrade already exists, don't add persist `jobDetail` if it already exists

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -493,7 +583,7 @@ private void upgradeJobs() throws Exception {
   }
 
-  private void upgradeTriggers() throws Exception {
+  private void upgradeTriggers(Table table) {
```
also need to check whether the trigger after upgrade already exists, don't add persist trigger if it already exists

gokulavasan replied:
Made that change.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -363,25 +395,99 @@ String removeAppVersion(String scheduleId) {
+          sleepTimeInMins.set(0);
+        } else if (e instanceof TransactionNotInProgressException) {
+          int currMaxRows = maxRows.get();
+          if (currMaxRows > 1) {
```
Just to be safe, let's keep the min to be 500. We don't want to upgrade one row at a time.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -273,6 +304,21 @@ private void removeTrigger(Table table, TriggerKey key) {
     table.delete(TRIGGER_KEY, col);
   }
 
+  private JobDetail readJob(JobKey jobKey) {
```
Is this used anywhere?

gokulavasan replied:
Yes, while trying to upgrade, we need to check if the upgraded job key already exists. If it exists, we don't replace it.

poornachandra reviewed (commented):
Few more comments...

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/StandaloneAppFabricServer.java:
```diff
@@ -70,11 +70,12 @@ public StandaloneAppFabricServer(CConfiguration cConf,
           metricsCollectionService, programRuntimeService, applicationLifecycleService,
           programLifecycleService, streamCoordinatorClient, servicesNames, handlerHookNames, namespaceAdmin,
-          systemArtifactLoader, pluginService, privilegesFetcherProxyService, routeStore);
+          systemArtifactLoader, pluginService, privilegesFetcherProxyService, appVersionUpgradeService, routeStore);
```
`StandaloneAppFabricServer` is used only in standalone mode, do we need `appVersionUpgradeService` here?

gokulavasan replied:
We don't technically need it. But this extends AppFabricServer. We need to pass in null and then don't start appVersionUpgradeService if is null etc. And we need to figure out whether it is sdk/distributed for the upgradehttphandler etc. Let me know if we need a different behavior for both. In SDK, the threads will quickly complete anyway.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:
```diff
@@ -98,9 +102,10 @@ public RunId apply(RunRecordMeta runRecordMeta) {
+  public AppMetadataStore(Table table, CConfiguration cConf, AtomicBoolean upgradeComplete) {
     super(table);
     this.cConf = cConf;
+    this.upgradeComplete = upgradeComplete;
```
I don't see this being used anywhere

gokulavasan replied:
Added uses of this now

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -88,16 +93,21 @@
-
   private static final Logger LOG = LoggerFactory.getLogger(DefaultStore.class);
   private static final DatasetId APP_META_INSTANCE_ID = NamespaceId.SYSTEM.dataset(Constants.AppMetaStore.TABLE);
+  private static final byte[] APP_VERSION_UPGRADE_KEY = Bytes.toBytes("upgrade.appmeta.appversion.complete");
```
Good to change this to be consistent with other such names

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -119,6 +131,16 @@ public DefaultStore(CConfiguration conf, DatasetFramework framework, Transaction
       RetryStrategies.retryOnConflict(20, 100)
     );
+    this.appVerUpgrade = new AtomicBoolean(false);
+    this.executorService = Executors.newSingleThreadExecutor(
```
Again, we can start this only if upgrade is not complete.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -881,14 +903,81 @@ void clear() throws Exception {
+          sleepTimeInMins.set(0);
+        } else if (e instanceof TransactionNotInProgressException) {
+          int currMaxRows = maxRows.get();
+          if (currMaxRows > 1) {
```
Same note about having min as 500

poornachandra reviewed (commented):
Few more comments...

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -363,25 +404,76 @@ String removeAppVersion(String scheduleId) {
+    final AtomicInteger sleepTimeInMins = new AtomicInteger(1);
+    LOG.info("Starting upgrade of StreamSizeScheduleStore.");
+    while (!isUpgradeComplete()) {
+      sleepTimeInMins.set(1);
```
Would be good to have this sleep at the end of the loop, no?

gokulavasan replied:
This is just resetting the sleep value to 1min.

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -227,7 +269,6 @@ public synchronized void delete(final ProgramId programId, final SchedulableProg
           String rowKey = getRowKey(programId, programType, scheduleName);
-
           String versionLessRowKey = removeAppVersion(rowKey);
           if (versionLessRowKey != null) {
```
We can add check for upgrade complete in this class too, right?

gokulavasan replied:
Done

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -81,24 +89,58 @@
+    checkUpgradeComplete(appVerUpgrade);
+    if (!appVerUpgrade.get()) {
+      executorService = Executors.newSingleThreadExecutor(
+        Threads.createDaemonThreadFactory("streamsize-schedulestore-appver-upgrade-checker"));
```
If we use a LoadingCache with a timeout to check for upgrade column in HBase table in class `UpgradeCompleteCheck`, we don't need to start upgrade checker threads. 

Until the upgrade is complete,  the loading cache will check the HBase table every timeout time to see if upgrade is complete. All other times it will return the cached value. Once the upgrade is done, a flag is set and the loading cache is no longer consulted.

poornachandra reviewed (commented):
A couple early comments...

poornachandra commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/MetadataStoreDataset.java:
```diff
@@ -365,6 +360,14 @@ public void deleteAll(MDSKey id, @Nullable Predicate<MDSKey> filter) {
 
+  public void delete(MDSKey id) {
+    try {
+      table.delete(id.getKey(), COLUMN);
```
Any specific reason this delete was changed to column delete vs the row delete that was being done earlier?

gokulavasan replied:
No specific reason, looks like all operations were being performed on that column. So just maintained that for uniformity.

gokulavasan commented:
@poornachandra Addressed all comments and tested the upgrade of records by background threads and the upgrade status endpoint. Please take a look again. Thanks!

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -397,18 +467,28 @@ private void upgradeVersionKeys() {
                      "format 'streamSizeSchedule:namespace:application:type:program:schedule'", oldRowKeyString);
```
Better to log with LIMITED_LOG

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -363,25 +381,77 @@ String removeAppVersion(String scheduleId) {
+    final AtomicInteger maxNumberUpdateRows = new AtomicInteger(1000);
+    final AtomicInteger sleepTimeInMins = new AtomicInteger(1);
+    LOG.info("Starting upgrade of {}.", NAME);
+    while (!isUpgradeComplete()) {
```
can have a local variable `final AtomicBoolean upgradeComplete` to be set true directly after `table.put(APP_VERSION_UPGRADE_KEY, COLUMN, Bytes.toBytes(ProjectInfo.getVersion().toString()));` and check the local variable `upgradeComplete` here instead of calling `isUpgradeComplete()`, which may not be updated immediately

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java:
```diff
@@ -360,22 +396,65 @@ public void apply() throws Exception {
+
+    final AtomicInteger sleepTimeInMins = new AtomicInteger(1);
+    LOG.info("Starting upgrade of {}.", NAME);
+    while (!isUpgradeComplete()) {
```
same comment as in stream size schedule store

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -363,25 +381,77 @@ String removeAppVersion(String scheduleId) {
+          int currMaxRows = maxNumberUpdateRows.get();
+          if (currMaxRows > 500) {
+            maxNumberUpdateRows.decrementAndGet();
+          }
```
we can stop retrying when `currMaxRows <= 500` or some other threshold is reached. Otherwise, this can go on forever, and block other transactions

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java:
```diff
@@ -363,25 +381,77 @@ String removeAppVersion(String scheduleId) {
+          if (currMaxRows > 500) {
+            maxNumberUpdateRows.decrementAndGet();
+          }
+          sleepTimeInMins.set(0);
```
Better to sleep for some time. Otherwise this will keep retrying and prevent other transaction from happening

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -881,14 +912,71 @@ void clear() throws Exception {
+          int currMaxRows = maxRows.get();
+          if (currMaxRows > 500) {
+            maxRows.decrementAndGet();
+          }
```
same comment as in `DatasetBasedStreamSizeScheduleStore`

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -881,14 +912,71 @@ void clear() throws Exception {
+          if (currMaxRows > 500) {
+            maxRows.decrementAndGet();
+          }
+          sleepTimeInMins.set(0);
```
same comment as in `DatasetBasedStreamSizeScheduleStore`

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -881,14 +912,71 @@ void clear() throws Exception {
+    // Repeated calls to upgradeComplete are necessary since it will trigger the cache to load the value from the table
+    // and that will eventually set the upgradeComplete flag to true, which is used by the methods in AppMetadataStore
+    // to check whether they need to do additional scans to accommodate old data formats.
+    while (!isUpgradeComplete()) {
```
same comment as in `DatasetBasedStreamSizeScheduleStore`

maochf commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java:
```diff
@@ -848,11 +877,12 @@ public boolean apply(RunRecordMeta runRecordMeta) {
+    upgradeDone &= upgradeVersionKeys(TYPE_RUN_RECORD_COMPLETED, RunRecordMeta.class, maxRows);
+    upgradeDone &= upgradeVersionKeys(TYPE_WORKFLOW_NODE_STATE, WorkflowNodeStateDetail.class, maxRows);
+    upgradeDone &= upgradeVersionKeys(TYPE_WORKFLOW_TOKEN, BasicWorkflowToken.class, maxRows);
+    return upgradeDone;
```
`this.upgradeComplete` is only updated in `DefaultStore.DefaultStoreUpgradeCacheLoader.load(byte[] key)`, otherwise, it's not updated. It can actually be updated here if `upgradeDone` is true. If this is done, maybe `DefaultStore.DefaultStoreUpgradeCacheLoader` won't be necessary and `DefaultStore.isUpgradeComplete()` can just return  `upgradeComplete` in AppMetadataStore

ecfm reviewed (commented):
Also found several calls of method `getVersionLessProgramKeyBuilder`, which may not be necessary when upgrade is complete in AppMetadataStore, please check the usages and add this condition if necessary. This also applies to the duplicate class `co.cask.cdap.logging.gateway.handlers.store.AppMetadataStore`

poornachandra commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AppVersionUpgradeService.java:
```diff
@@ -0,0 +1,106 @@
+  private final DefaultStore defaultStore;
+
+  @Inject
+  AppVersionUpgradeService(DatasetBasedStreamSizeScheduleStore streamSizeScheduleStore,
```
`AppVersionUpgradeService` can be converted into a `RetryOnStartFailureService`, and while starting the upgrade service you can check if the stores have already been upgraded. If there are any exceptions the upgrade service will restart. Only when the checks return false, the upgrade jobs can be submitted. 

This will also remove the need for store initialized flags that are being used in all stores.

ecfm commented:
Since `DatasetBasedStreamSizeScheduleStore` and `DatasetBasedTimeScheduleStore` are actually both singleton, they can actually have an AtomicBoolean as isUpgradeComplete flag. The flag is updated at startup by reading from the table and updated when upgrade is complete. No cache loader will be needed and table reading will only be done once at startup.

poornachandra reviewed (commented):
@gokulavasan Just one comment as of now

ecfm commented:
need to remove the extra space in "No meta for started run record for namespace %s app %s version % s program type %s " in https://github.com/caskdata/cdap/blob/8d8d3a352f8ea64a75280c86bce74b527419ceb7/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java#L398

ecfm commented:
There are actually more operations in `AppMetadataStore` where `isUpgradeComplete` should be checked but not, and they don't call `getVersionLessProgramKeyBuilder` but call `MDSKey.Builder()` directly (which is not a good idea). For instance, `getApplication(String namespaceId, String appId, String versionId)`, `writeApplication(String namespaceId, String appId, String versionId, ApplicationSpecification spec)`, `deleteApplication(String namespaceId, String appId, String versionId)`,  `updateAppSpec(String namespaceId, String appId, String versionId, ApplicationSpecification spec)` are all like this. Accepting parameters with ApplicationId or ProgramId parts as String is not a good idea. We should pass in Id objects and call methods like `getProgramKeyBuilder` to construct MDSKey

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -881,14 +912,71 @@ void clear() throws Exception {
+            boolean upgradeComplete = store.upgradeVersionKeys(maxRows.get());
+            if (upgradeComplete) {
+              store.setUpgradeComplete(APP_VERSION_UPGRADE_KEY);
+            }
```
why not set `this.upgradeComplete` to true here?

ecfm commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -89,26 +98,35 @@
   private final DatasetFramework dsFramework;
   private final Transactional transactional;
 
+  private static LoadingCache<byte[], Boolean> upgradeCacheLoader;
```
can directly initialize it here instead of in the constructor, and no need of `cacheLoaderInitialized`

ecfm commented:
LGTM
